core/src/blocktree/db.rs

```diff
@@ -7,129 +6,398 @@ use serde::de::DeserializeOwned;
-        Ok(())
+    pub fn seek_to_first(&mut self) {
+        self.db_cursor.seek_to_first();
     }
 
-    fn raw_iterator(&self) -> D::Cursor {
-        let db = self.db();
-        db.raw_iterator_cf(self.handle())
-            .expect("Expected to be able to open database iterator")
+    pub fn next(&mut self) {
+        self.db_cursor.next();
+    }
+
+    pub fn key(&self) -> Option<C::Index> {
+        if let Some(key) = self.db_cursor.key() {
+            Some(C::index(key.borrow()))
+        } else {
+            None
+        }
     }
 
-    fn handle(&self) -> D::ColumnFamily;
+    pub fn value_bytes(&self) -> Option<Vec<u8>> {
+        self.db_cursor.value()
+    }
+}
 
-    fn db(&self) -> &Arc<D>;
+impl<B, C> Cursor<B, C>
+where
+    B: Backend,
+    C: TypedColumn<B>,
+{
+    pub fn value(&self) -> Option<C::Type> {
+        if let Some(bytes) = self.db_cursor.value() {
+            let value = deserialize(&bytes).ok()?;
+            Some(value)
+        } else {
+            None
+        }
+    }
 }
 
-pub trait IndexColumn<D: Database>: LedgerColumnFamilyRaw<D> {
-    type Index;
+impl<B, C> LedgerColumn<B, C>
+where
+    B: Backend,
+    C: Column<B>,
+{
+    pub fn new(db: &Arc<Database<B>>) -> Self {
+        LedgerColumn {
+            db: Arc::clone(db),
+            column: PhantomData,
+        }
+    }
 
-    fn get_by_index(&self, index: &Self::Index) -> Result<Option<Vec<u8>>> {
-        let db = self.db();
-        let data_bytes = db.get_cf(self.handle(), Self::key(index).borrow())?;
-        Ok(data_bytes.map(|x| x.to_vec()))
+    pub fn put_bytes(&self, key: C::Index, value: &[u8]) -> Result<()> {
+        self.db
+            .backend
+            .put_cf(self.handle(), C::key(key).borrow(), value)
     }
 
-    fn put_by_index(&self, index: &Self::Index, serialized_value: &[u8]) -> Result<()> {
-        let db = self.db();
-        db.put_cf(self.handle(), Self::key(index).borrow(), &serialized_value)?;
-        Ok(())
+    pub fn get_bytes(&self, key: C::Index) -> Result<Option<Vec<u8>>> {
+        self.db.backend.get_cf(self.handle(), C::key(key).borrow())
     }
 
-    fn delete_by_index(&self, index: &Self::Index) -> Result<()> {
-        self.delete(Self::key(index).borrow())
+    pub fn delete(&self, key: C::Index) -> Result<()> {
+        self.db
+            .backend
+            .delete_cf(self.handle(), C::key(key).borrow())
     }
 
-    fn index(key: &D::Key) -> Self::Index;
+    pub fn cursor(&self) -> Result<Cursor<B, C>> {
+        self.db.cursor()
+    }
+
+    pub fn iter(&self) -> Result<impl Iterator<Item = (C::Index, Vec<u8>)>> {
+        self.db.iter::<C>()
+    }
+
+    pub fn handle(&self) -> B::ColumnFamily {
+        self.db.cf_handle::<C>()
+    }
+
+    pub fn is_empty(&self) -> Result<bool> {
+        let mut cursor = self.cursor()?;
+        cursor.seek_to_first();
+        Ok(!cursor.valid())
+    }
+}
+
+impl<B, C> LedgerColumn<B, C>
+where
+    B: Backend,
+    C: TypedColumn<B>,
```

Comment:
This is a clever way to overload the LedgerColumn with with a bit of type safety 👯